hail/python/hail/methods/relatedness/identity_by_descent.py

```diff
@@ -87,18 +91,138 @@ def identity_by_descent(dataset, maf=None, bounded=True, min=None, max=None) ->
+            'bounded': bounded,
+            'min': min,
+            'max': max,
+        })).persist()
+
+    min = min or 0
+    max = max or 1
+    if not 0 <= min <= max <= 1:
+        raise Exception(f"invalid pi hat filters {min} {max}")
+
+    sample_ids = dataset.s.collect()
+    if len(sample_ids) != len(set(sample_ids)):
+        raise Exception('duplicate sample ids found')
+
+    dataset = dataset.annotate_entries(
+        n_alt_alleles=hl.or_else(dataset.GT.n_alt_alleles(), 0),
+        is_hom_ref=hl.or_else(dataset.GT.is_hom_ref(), 0),
+        is_het=hl.or_else(dataset.GT.is_het(), 0),
+        is_hom_var=hl.or_else(dataset.GT.is_hom_var(), 0),
+        is_missing=hl.is_missing(dataset.GT),
+        is_not_missing=hl.is_defined(dataset.GT),
+    )
+
+    T = 2 * hl.agg.count_where(hl.is_defined(dataset.GT))
+    X = hl.agg.sum(dataset.GT.n_alt_alleles())
+    Y = T - X
+
+    if maf is not None:
+        p = dataset.__maf
+    else:
+        p = X / T
+
+    q = 1 - p
+
+    dataset = dataset.annotate_rows(
+        _e00=(2 * (p ** 2) * (q ** 2) * ((X - 1) / X) * ((Y - 1) / Y) * (T / (T - 1)) * (T / (T - 2)) * (T / (T - 3))),
+        _e10=(4 * (p ** 3) * q * ((X - 1) / X) * ((X - 2) / X) * (T / (T - 1)) * (T / (T - 2)) * (T / (T - 3))
+              + 4 * p * (q ** 3) * ((Y - 1) / X) * ((Y - 2) / X) * (T / (T - 1)) * (T / (T - 2)) * (T / (T - 3))),
+        _e20=((p ** 4) * ((X - 1) / X) * ((X - 2) / X) * ((X - 3) / X) * (T / (T - 1)) * (T / (T - 2)) * (T / (T - 3))
+              + (q ** 4) * ((Y - 1) / Y) * ((Y - 2) / Y) * ((Y - 3) / Y) * (T / (T - 1)) * (T / (T - 2)) * (T / (T - 3))
+              + 4 * (p ** 2) * (q ** 2) * ((X - 1) / X) * ((Y - 1) / Y) * (T / (T - 1)) * (T / (T - 2)) * (T / (T - 3))),
+        _e11=(2 * (p ** 2) * q * ((X - 1) / X) * (T / (T - 1)) * (T / (T - 2))
+              + 2 * p * (q ** 2) * ((Y - 1) / Y) * (T / (T - 1)) * (T / (T - 2))),
+        _e21=((p ** 3) * ((X - 1) / X) * ((X - 2) / X) * (T / (T - 1)) * (T / (T - 2))
+              + (q ** 3) * ((Y - 1) / Y) * ((Y - 2) / Y) * (T / (T - 1)) * (T / (T - 2))
+              + (p ** 2) * q * ((X - 1) / X) * (T / (T - 1)) * (T / (T - 2))
+              + p * (q ** 2) * ((Y - 1) / Y) * (T / (T - 1)) * (T / (T - 2))),
+        _e22=(T / 2)
+    )
+
+    expectations = dataset.aggregate_rows(hl.struct(
+        e00=hl.agg.sum(dataset._e00),
+        e10=hl.agg.sum(dataset._e10),
+        e20=hl.agg.sum(dataset._e20),
+        e11=hl.agg.sum(dataset._e11),
+        e21=hl.agg.sum(dataset._e21),
+        e22=hl.agg.sum(dataset._e22)
+    ))
+
+    IS_HOM_REF = BlockMatrix.from_entry_expr(dataset.is_hom_ref).checkpoint(hl.utils.new_temp_file())
+    IS_HET = BlockMatrix.from_entry_expr(dataset.is_het).checkpoint(hl.utils.new_temp_file())
+    IS_HOM_VAR = BlockMatrix.from_entry_expr(dataset.is_hom_var).checkpoint(hl.utils.new_temp_file())
+    NOT_MISSING = IS_HOM_REF + IS_HET + IS_HOM_VAR
+
+    total_possible_ibs = NOT_MISSING.T @ NOT_MISSING
```

Comment:
Not sure if checkpointing NOT_MISSING before the matmul if worth the cost of the write. Might try benchmarking both ways.